platform/linux-generic/pktio/socket_mmap.c

```diff
@@ -251,6 +263,50 @@ static inline unsigned pkt_mmap_v2_rx(pktio_entry_t *pktio_entry,
 	return nb_rx;
 }
 
+static unsigned handle_pending_frames(int sock, struct ring *ring, int frames)
+{
+	int i;
+	int retry = 0;
+	unsigned nb_tx = 0;
+	unsigned frame_num;
+	unsigned frame_count = ring->rd_num;
+	unsigned first_frame_num = ring->frame_num;
+
+	for (frame_num = first_frame_num, i = 0; i < frames; i++) {
+		struct tpacket2_hdr *hdr = ring->rd[frame_num].iov_base;
+
+		if (odp_likely(hdr->tp_status == TP_STATUS_AVAILABLE ||
+			       hdr->tp_status == TP_STATUS_SENDING)) {
+			nb_tx++;
+		} else if (hdr->tp_status == TP_STATUS_SEND_REQUEST) {
+			if (retry++ < TX_RETRIES) {
+				struct timespec ts = { .tv_nsec = TX_RETRY_NSEC,
+						       .tv_sec = 0 };
+
+				sendto(sock, NULL, 0, MSG_DONTWAIT, NULL, 0);
+				nanosleep(&ts, NULL);
+				i--;
+				continue;
+			} else {
+				hdr->tp_status = TP_STATUS_AVAILABLE;
+			}
+		} else { /* TP_STATUS_WRONG_FORMAT */
+			/* Don't try re-sending frames after failure */
+			for (; i < frames; i++) {
+				hdr = ring->rd[frame_num].iov_base;
+				hdr->tp_status = TP_STATUS_AVAILABLE;
+				frame_num = next_frame(frame_num, frame_count);
+			}
+			break;
+		}
+		frame_num = next_frame(frame_num, frame_count);
+	}
+
+	ring->frame_num = (first_frame_num + nb_tx) % frame_count;
```

Comment:
`ring->frame_num = next_frame(first_frame_num + nb_tx - 1, frame_count);` is an alternative here.

Reply:
Will fix.